src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +190,344 @@ public void emit(Event event)
+      int bufferOffset = batchingStrategy.writeBatchStart(buffer);
+      System.arraycopy(eventBytes, 0, buffer, bufferOffset, eventBytes.length);
+      bufferOffset += eventBytes.length;
+      bufferOffset = batchingStrategy.writeBatchEnd(buffer, bufferOffset);
       try {
-        serializeTo(messages, baos);
-        return baos.toByteArray();
+        sendWithRetries(buffer, bufferOffset, 1);
       }
-      catch (IOException e) {
-        // There's no reason to have IOException in the signature of this method, since BAOS won't throw them.
-        throw Throwables.propagate(e);
+      finally {
+        buffersToReuse.add(buffer);
       }
     }
 
-    /**
-     * Serialize and compress (gzip) messages into a batch.
-     *
-     * @param messages list of JSON objects, one per message
-     * @return serialized and compressed JSON array
-     */
-    private byte[] serializeAndCompressBatch(List<byte[]> messages) throws IOException
+    private void sendWithRetries(final byte[] buffer, final int length, final int eventCount)
     {
-      final ByteArrayOutputStream baos = new ByteArrayOutputStream();
-      try (
-          final GZIPOutputStream gzos = new GZIPOutputStream(baos)
-      ) {
-        serializeTo(messages, gzos);
+      try {
+        RetryUtils.retry(
+            new Callable<Void>()
+            {
+              @Override
+              public Void call() throws Exception
+              {
+                send(buffer, length);
+                return null;
+              }
+            },
+            Predicates.<Throwable>alwaysTrue(),
+            MAX_SEND_RETRIES
+        );
+        totalEmittedEvents.addAndGet(eventCount);
+      }
+      catch (Exception e) {
+        log.error(e, "Failed to send events to url[%s]", config.getRecipientBaseUrl());
       }
-      return baos.toByteArray();
     }
 
-    /**
-     * Serialize messages into a batch
-     *
-     * @param messages list of JSON objects, one per message
-     * @param os output stream with serialized batch
-     * @throws IOException if an I/O error occurs.
-     */
-    private void serializeTo(List<byte[]> messages, OutputStream os) throws IOException
+    private void send(byte[] buffer, int length) throws Exception
     {
-      boolean first = true;
-      os.write(config.getBatchingStrategy().batchStart());
-      for (final byte[] message : messages) {
-        if (first) {
-          first = false;
-        } else {
-          os.write(config.getBatchingStrategy().messageSeparator());
+      final Request request = new Request(HttpMethod.POST, url);
+      byte[] payload;
+      int payloadLength;
+      ContentEncoding contentEncoding = config.getContentEncoding();
+      if (contentEncoding != null) {
+        switch (contentEncoding) {
+          case GZIP:
+            try (GZIPOutputStream gzipOutputStream = acquireGzipOutputStream(length)) {
+              gzipOutputStream.write(buffer, 0, length);
```

Comment:
older versions of java swallow flush exceptions on close for FilterOutputStream like gzip, but I think they fixed it in java8